File: dynamic-programming/coin-change/coin-change.ts
function coinChange(coins: number[], amount: number): number {
  const dp: number[] = new Array(amount + 1).fill(~~((1 << 31) - 2));
  let n: number = dp.length;
  let m: number = dp.length;

  dp[0] = 0;

  for (let i = 0; i < m; i++) {
    let coin: number = coins[i];
    for (let j = coin; j < n; j++) {
      dp[j] = Math.min(dp[j], dp[j - coin] + 1);
    }
  }

  if (dp[amount] == ~~((1 << 31) - 2)) {
    return -1;
  }

  return dp[amount];
}

File: dynamic-programming/coin-change/coin-change.java
class Solution {
  public int coinChange(int[] coins, int amount) {
    int[] dp = new int[amount + 1];
    int n = dp.length;
    int m = coins.length;
    Arrays.fill(dp, Integer.MAX_VALUE - 1);

    dp[0] = 0;

    for (int i = 0; i < m; i++) {
      int coin = coins[i];
      for (int j = coin; j < n; j++) {
        dp[j] = Math.min(dp[j], dp[j - coin] + 1);
      }
    }

    if (dp[n - 1] == Integer.MAX_VALUE - 1) {
      return -1;
    }

    return dp[n - 1];

  }
}
